R/credentials_gce.R

```diff
@@ -114,36 +138,112 @@ credentials_gce <- function(scopes = "https://www.googleapis.com/auth/cloud-plat
 #' @keywords internal
 #' @export
 GceToken <- R6::R6Class("GceToken", inherit = httr::Token2.0, list(
-  #' @description Print token
-  print = function(...) {
-    cat("<GceToken>")
+  #' @description Get an access for a GCE service account.
+  #' @param params A list of parameters for `fetch_gce_access_token()`.
+  #' @return A GceToken.
+  initialize = function(params) {
+    gargle_debug("GceToken initialize")
+    self$params <- params
+    self$init_credentials()
   },
-  #' @description Placeholder implementation of required method
+  #' @description Request an access token.
   init_credentials = function() {
-    self$credentials <- list(access_token = NULL)
+    gargle_debug("GceToken init_credentials")
+    token <- fetch_gce_access_token(
+      self$params$scope,
+      service_account = self$params$service_account
+    )
+
+    # find out the scopes actually obtained
+    # https://www.googleapis.com/oauth2/v3/tokeninfo
+    req <- request_build(
+      method = "GET",
+      path = "oauth2/v3/tokeninfo",
+      params = list(access_token = token$access_token),
+      base_url = "https://www.googleapis.com"
+    )
+    resp <- request_make(req)
+    info <- response_process(resp)
+    actual_scopes <- normalize_scopes(strsplit(info$scope, split = "\\s+")[[1]])
+
+    missing <- setdiff(self$params$scope, actual_scopes)
+    if (length(missing) > 0) {
+      gargle_debug(c(
+        "!" = "{cli::qty(length(missing))}{?This/These} requested \\
+             scope{?s} {?is/are} not among the scopes for the \\
+             access token returned by the metadata server:",
+        bulletize(missing, bullet = "x"),
+        "i" = "If there are problems downstream, this might be the root cause."
+      ))
+      gargle_debug(c(
+        "!" = "Setting token scopes to:",
+        bulletize(actual_scopes)
+      ))
+      self$params$scope <-actual_scopes
```

Comment:
This is new-ish, in the sense that I'm recording the token's actual scopes vs. recording the requested scopes. Of course, the actual scopes were always whatever they were, even if it's not what we asked for. With this change, we're just being more honest with ourselves about what the token we just got can actually do. This will affect a `$refresh()` call, i.e. these will be the requested scopes upon refresh.

Reply:
What this looks like in the debugging output:

```
! Setting token scopes to:
• https://www.googleapis.com/auth/cloud-platform
```